src/jdk.incubator.jextract/share/classes/jdk/internal/jextract/impl/AnnotationWriter.java

```diff
@@ -0,0 +1,126 @@
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package jdk.internal.jextract.impl;
+
+/*
+ * Copyright (c) 2020, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+import jdk.incubator.jextract.Type;
+import java.util.stream.Collectors;
+
+public class AnnotationWriter implements Type.Visitor<String, Void> {
+    @Override
+    public String visitPrimitive(Type.Primitive t, Void aVoid) {
+        if (t.kind().layout().isEmpty()) {
+            return "void"; //skip for now
+        } else {
+            return t.kind().typeName();
+        }
+    }
+
+    @Override
+    public String visitDelegated(Type.Delegated t, Void aVoid) {
+        if (t.kind() == Type.Delegated.Kind.TYPEDEF) {
+            return t.name().get();
+        } else if (t.kind() == Type.Delegated.Kind.POINTER) {
+            String typeStr = t.type().accept(this, null);
+            if (t.type() instanceof Type.Function) {
+                return typeStr.replace("(", "(*)(");
```

Comment:
This might be problematic for multiple nested function types, e.g.

```Java
void (int, void (*)(void)) -> void (*)(int, void (*)(*)(*)(void))
```

The return type could also be a function pointer, so it's not always the first occurrence that needs to be replaced either.

Reply:
Might just want to revisit the return type and argumentTypes, similar to what you do in visitFunction, but use `(*)(` as a joiner instead of `(`.